File: cli-sender.js
#!/usr/bin/env node

/**
 * Simple SMS Sender using fetch (Node.js version)
 * Based on cli-sender.php
 * 
 * Usage: node cli-sender.js --mobile=<number> --message=<text> [--api-key=<key>] [--api-url=<url>]
 * 
 * Example: node cli-sender.js --mobile=639176738989 --message="Hello World"
 */

// Default configuration
const API_KEY = 'sign-up-and-create-api-key';
const API_BASE_URL = 'https://www.etherney.org';
const API_SEND_URL = API_BASE_URL + '/api/send-sms.php';

/**
 * Parse command line arguments
 */
function parseArgs(argv) {
    const options = {
        mobile: null,
        message: null,
        apiKey: null,
        apiUrl: null,
        help: false
    };

    for (let i = 2; i < argv.length; i++) {
        const arg = argv[i];
        if (arg === '--help' || arg === '-h') {
            options.help = true;
        } else if (arg.startsWith('--mobile=')) {
            options.mobile = arg.substring(9);
        } else if (arg.startsWith('--message=')) {
            options.message = arg.substring(10);
        } else if (arg.startsWith('--api-key=')) {
            options.apiKey = arg.substring(10);
        } else if (arg.startsWith('--api-url=')) {
            options.apiUrl = arg.substring(10);
        }
    }

    return options;
}

/**
 * Display help message
 */
function displayHelp() {
    console.log("Simple SMS Sender using fetch (Node.js)");
    console.log("=======================================\n");
    console.log("Usage: node cli-sender.js [options]\n");
    console.log("Options:");
    console.log("  --mobile=<number>      Mobile number (without +)");
    console.log("  --message=<text>       Message text to send");
    console.log("  --api-key=<key>        API key (optional, uses default if not provided)");
    console.log("  --api-url=<url>        API base URL (optional, default: https://www.etherney.org)");
    console.log("  --help                 Show this help message\n");
    console.log("Examples:");
    console.log("  node cli-sender.js --mobile=639176738989 --message=\"Hello\"");
    console.log("  node cli-sender.js --mobile=639176738989 --message=\"Test\" --api-key=your_key");
    console.log("  node cli-sender.js --mobile=639176738989 --message=\"Test\" --api-url=http://your-server.com");
}

/**
 * Validate mobile number format
 */
function validateMobileNumber(mobileNo) {
    const regex = /^[0-9]{10,15}$/;
    return regex.test(mobileNo);
}

/**
 * Send SMS using fetch
 */
async function sendSms(mobileNo, message, apiKey, apiUrl) {
    // Validate inputs
    if (!message || !message.trim()) {
        return { success: false, error: 'Message cannot be empty' };
    }

    if (!validateMobileNumber(mobileNo)) {
        return { success: false, error: 'Invalid mobile number format. Use 10-15 digits without +' };
    }

    // Prepare data
    const data = new URLSearchParams();
    data.append('api_key', apiKey);
    data.append('mobile_no', mobileNo);
    data.append('message', message);

    try {
        const response = await fetch(apiUrl, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/x-www-form-urlencoded'
            },
            body: data.toString(),
            // Disable SSL verification like PHP does
            agent: new (require('https').Agent)({ rejectUnauthorized: false })
        });

        const responseText = await response.text();
        
        let result;
        try {
            result = JSON.parse(responseText);
        } catch (e) {
            return { 
                success: false, 
                error: `Invalid JSON response: ${e.message}`,
                rawResponse: responseText
            };
        }

        return result;
    } catch (error) {
        return { 
            success: false, 
            error: `HTTP error: ${error.message}` 
        };
    }
}

/**
 * Format currency amount
 */
function formatCurrency(amount) {
    try {
        const num = parseFloat(amount);
        return `$${num.toFixed(4)}`;
    } catch (e) {
        return `$${amount}`;
    }
}

/**
 * Main execution
 */
async function main() {
    const options = parseArgs(process.argv);

    // Show help if requested or no arguments
    if (options.help || process.argv.length === 2) {
        displayHelp();
        return;
    }

    // Check required arguments
    if (!options.mobile || !options.message) {
        console.log("Error: Both --mobile and --message are required.\n");
        displayHelp();
        return;
    }

    // Set API key and URL
    const apiKey = options.apiKey || API_KEY;
    const apiUrl = options.apiUrl ? options.apiUrl + '/api/send-sms.php' : API_SEND_URL;

    // Display info
    console.log("Simple SMS Sender");
    console.log("================\n");
    console.log(`Mobile: ${options.mobile}`);
    const messagePreview = options.message.length > 50 
        ? options.message.substring(0, 50) + '...' 
        : options.message;
    console.log(`Message: ${messagePreview}`);
    console.log(`Message length: ${options.message.length} characters`);
    console.log(`API Key: ${apiKey.substring(0, 8)}...`);
    console.log(`API URL: ${apiUrl}\n`);

    // Send SMS
    console.log("Sending SMS...");
    const result = await sendSms(options.mobile, options.message, apiKey, apiUrl);

    // Display result
    if (result.success) {
        console.log("✓ SMS sent successfully!");
        const data = result.data || {};
        console.log(`  SMS ID: ${data.sms_id || 'N/A'}`);
        console.log(`  Status: ${data.status || 'N/A'}`);
        console.log(`  Provider: ${data.provider || 'N/A'}`);
        console.log(`  To: ${data.to || 'N/A'}`);

        if (data.balance_before !== undefined && data.balance_after !== undefined) {
            console.log(`  Credits before: ${formatCurrency(data.balance_before)}`);
            console.log(`  Credits after:  ${formatCurrency(data.balance_after)}`);
        }
    } else {
        console.log(`✗ Failed to send SMS: ${result.error || 'Unknown error'}`);

        const data = result.data || {};
        if (data.balance_before !== undefined && data.balance_after !== undefined) {
            console.log(`  Credits before: ${formatCurrency(data.balance_before)}`);
            console.log(`  Credits after:  ${formatCurrency(data.balance_after)}`);
        }
    }
}

// Run main function
main().catch(error => {
    console.error(`Unexpected error: ${error.message}`);
    process.exit(1);
});

File: cli-sender.py
#!/usr/bin/env python3
"""
Simple SMS Sender using HTTP requests
Python version of cli-sender.php

Usage: python cli-sender.py --mobile=<number> --message=<text> [--api-key=<key>] [--api-url=<url>]

Example: python cli-sender.py --mobile=639176738989 --message="Hello World"
"""

import sys
import re
import json
import argparse
from urllib.parse import urlencode

try:
    import requests
    HAS_REQUESTS = True
except ImportError:
    HAS_REQUESTS = False
    import urllib.request
    import urllib.error

# Default configuration (same as PHP)
API_KEY = 'sign-up-and-create-api-key'
API_BASE_URL = 'https://www.etherney.org'
API_SEND_URL = API_BASE_URL + '/api/send-sms.php'


def parse_args():
    """Parse command line arguments."""
    parser = argparse.ArgumentParser(
        description='Simple SMS Sender using HTTP requests',
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
Examples:
  %(prog)s --mobile=639176738989 --message="Hello"
  %(prog)s --mobile=639176738989 --message="Test" --api-key=your_key
  %(prog)s --mobile=639176738989 --message="Test" --api-url=http://your-server.com
        """
    )
    parser.add_argument('--mobile', required=True, help='Mobile number (without +)')
    parser.add_argument('--message', required=True, help='Message text to send')
    parser.add_argument('--api-key', help='API key (optional, uses default if not provided)')
    parser.add_argument('--api-url', help='API base URL (optional, default: https://www.etherney.org)')
    
    # For compatibility with PHP script's --help behavior
    if len(sys.argv) == 1:
        parser.print_help()
        sys.exit(0)
    
    return parser.parse_args()


def send_sms_requests(mobile_no, message, api_key, api_url):
    """Send SMS using requests library."""
    data = {
        'api_key': api_key,
        'mobile_no': mobile_no,
        'message': message
    }
    
    try:
        response = requests.post(
            api_url,
            data=data,
            headers={'Content-Type': 'application/x-www-form-urlencoded'},
            timeout=30,
            verify=False  # Disable SSL verification like PHP does
        )
        response.raise_for_status()
        return response.json()
    except requests.exceptions.RequestException as e:
        return {'success': False, 'error': f'HTTP error: {str(e)}'}
    except json.JSONDecodeError as e:
        return {'success': False, 'error': f'Invalid JSON response: {str(e)}'}


def send_sms_urllib(mobile_no, message, api_key, api_url):
    """Send SMS using urllib (fallback when requests not available)."""
    data = urlencode({
        'api_key': api_key,
        'mobile_no': mobile_no,
        'message': message
    }).encode('utf-8')
    
    req = urllib.request.Request(
        api_url,
        data=data,
        headers={'Content-Type': 'application/x-www-form-urlencoded'},
        method='POST'
    )
    
    # Disable SSL verification (not straightforward with urllib)
    import ssl
    context = ssl._create_unverified_context()
    
    try:
        with urllib.request.urlopen(req, timeout=30, context=context) as response:
            response_text = response.read().decode('utf-8')
            return json.loads(response_text)
    except urllib.error.URLError as e:
        return {'success': False, 'error': f'URL error: {str(e)}'}
    except json.JSONDecodeError as e:
        return {'success': False, 'error': f'Invalid JSON response: {str(e)}'}
    except Exception as e:
        return {'success': False, 'error': f'Unexpected error: {str(e)}'}


def send_sms(mobile_no, message, api_key, api_url):
    """Send SMS with validation and appropriate HTTP client."""
    # Validate inputs
    if not message or not message.strip():
        return {'success': False, 'error': 'Message cannot be empty'}
    
    if not re.match(r'^[0-9]{10,15}$', mobile_no):
        return {'success': False, 'error': 'Invalid mobile number format. Use 10-15 digits without +'}
    
    # Send using requests if available, otherwise urllib
    if HAS_REQUESTS:
        return send_sms_requests(mobile_no, message, api_key, api_url)
    else:
        return send_sms_urllib(mobile_no, message, api_key, api_url)


def format_currency(amount):
    """Format amount as currency with 4 decimal places."""
    try:
        return f"${float(amount):.4f}"
    except (ValueError, TypeError):
        return f"${amount}"


def main():
    """Main execution function."""
    args = parse_args()
    
    # Set API key and URL
    api_key = args.api_key if args.api_key else API_KEY
    api_url = args.api_url + '/api/send-sms.php' if args.api_url else API_SEND_URL
    
    # Display info
    print("Simple SMS Sender")
    print("================\n")
    print(f"Mobile: {args.mobile}")
    message_preview = args.message[:50] + ('...' if len(args.message) > 50 else '')
    print(f"Message: {message_preview}")
    print(f"Message length: {len(args.message)} characters")
    print(f"API Key: {api_key[:8]}...")
    print(f"API URL: {api_url}\n")
    
    # Send SMS
    print("Sending SMS...")
    result = send_sms(args.mobile, args.message, api_key, api_url)
    
    # Display result
    if result.get('success'):
        print("✓ SMS sent successfully!")
        data = result.get('data', {})
        print(f"  SMS ID: {data.get('sms_id', 'N/A')}")
        print(f"  Status: {data.get('status', 'N/A')}")
        print(f"  Provider: {data.get('provider', 'N/A')}")
        print(f"  To: {data.get('to', 'N/A')}")
        
        if 'balance_before' in data and 'balance_after' in data:
            print(f"  Credits before: {format_currency(data['balance_before'])}")
            print(f"  Credits after:  {format_currency(data['balance_after'])}")
    else:
        print(f"✗ Failed to send SMS: {result.get('error', 'Unknown error')}")
        
        data = result.get('data', {})
        if 'balance_before' in data and 'balance_after' in data:
            print(f"  Credits before: {format_currency(data['balance_before'])}")
            print(f"  Credits after:  {format_currency(data['balance_after'])}")


if __name__ == '__main__':
    main()